LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/DefaultLoRaDataRequestHandler.cs

```diff
@@ -152,71 +160,89 @@ public async Task<LoRaDeviceRequestProcessResult> ProcessRequestAsync(LoRaReques
+                                    {
+                                        return new LoRaDeviceRequestProcessResult(loRaDevice, request, LoRaDeviceRequestFailedReason.HandledByAnotherGateway);
+                                    }
+
+                                    requiresConfirmation = true;
+                                }
+                            }
+                            else
                             {
                                 if (string.IsNullOrEmpty(loRaDevice.SensorDecoder))
                                 {
-                                    Logger.Log(loRaDevice.DevEUI, $"no decoder set in device twin. port: {fportUp}", LogLevel.Debug);
-                                    payloadData = Convert.ToBase64String(decryptedPayloadData);
+                                    Logger.Log(loRaDevice.DevEUI, $"no decoder set in device twin. port: {payloadPort}", LogLevel.Debug);
+                                    payloadData = new UndecodedPayload(decryptedPayloadData);
                                 }
                                 else
                                 {
-                                    Logger.Log(loRaDevice.DevEUI, $"decoding with: {loRaDevice.SensorDecoder} port: {fportUp}", LogLevel.Debug);
-                                    payloadData = await this.payloadDecoder.DecodeMessageAsync(decryptedPayloadData, fportUp, loRaDevice.SensorDecoder);
+                                    Logger.Log(loRaDevice.DevEUI, $"decoding with: {loRaDevice.SensorDecoder} port: {payloadPort}", LogLevel.Debug);
+                                    var decodePayloadResult = await this.payloadDecoder.DecodeMessageAsync(loRaDevice.DevEUI, decryptedPayloadData, payloadPort, loRaDevice.SensorDecoder);
+                                    payloadData = decodePayloadResult.GetDecodedPayload();
+
+                                    if (decodePayloadResult.CloudToDeviceMessage != null)
+                                    {
+                                        if (string.IsNullOrEmpty(decodePayloadResult.CloudToDeviceMessage.DevEUI) || string.Equals(loRaDevice.DevEUI, decodePayloadResult.CloudToDeviceMessage.DevEUI, StringComparison.InvariantCultureIgnoreCase))
+                                        {
+                                            // sending c2d to same device
+                                            cloudToDeviceMessage = decodePayloadResult.CloudToDeviceMessage;
+                                            fcntDown = await this.EnsureHasFcntDownAsync(loRaDevice, fcntDown, payloadFcnt, frameCounterStrategy);
+
+                                            if (!fcntDown.HasValue || fcntDown <= 0)
+                                            {
+                                                // We did not get a valid frame count down, therefore we should not process the message
+                                                _ = cloudToDeviceMessage.AbandonAsync();
```

Comment:
The intention is to not wait for it to be finished. So, afaik, it will return as soon as the amqp/mqtt will be sent. Could use `Task.Run` to make 100% in parallel.

Reply:
Ok. If you do not want to await, can't you just remove the _ = ? Is there a style rule complaining? I wouldn't start another thread for this.

Reply:
Discard was added to suppress warning [CS4014](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs4014).